rst_files/finite_markov.rst

```diff
@@ -1294,79 +1268,35 @@ compare it to the stationary probability.
 Exercise 2
 ----------
 
-First save the data into a file called ``web_graph_data.txt`` by
-executing the next cell
-
 .. code-block:: julia
 
-    f = open("web_graph_data.txt", "w")
-    contents = """a -> d;
-    a -> f;
-    b -> j;
-    b -> k;
-    b -> m;
-    c -> c;
-    c -> g;
-    c -> j;
-    c -> m;
-    d -> f;
-    d -> h;
-    d -> k;
-    e -> d;
-    e -> h;
-    e -> l;
-    f -> a;
-    f -> b;
-    f -> j;
-    f -> l;
-    g -> b;
-    g -> j;
-    h -> d;
-    h -> g;
-    h -> l;
-    h -> m;
-    i -> g;
-    i -> h;
-    i -> n;
-    j -> e;
-    j -> i;
-    j -> k;
-    k -> n;
-    l -> m;
-    m -> g;
-    n -> c;
-    n -> j;
-    n -> m;
-    """
-    write(f, contents)
-    close(f)
+    web_graph_data = sort(Dict('a' => ['d', 'f'],
+                               'b' => ['j', 'k', 'm'],
+                               'c' => ['c', 'g', 'j', 'm'],
+                               'd' => ['f', 'h', 'k'],
+                               'e' => ['d', 'h', 'l'],
+                               'f' => ['a', 'b', 'j', 'l'],
+                               'g' => ['b', 'j'],
+                               'h' => ['d', 'g', 'l', 'm'],
+                               'i' => ['g', 'h', 'n'],
+                               'j' => ['e', 'i', 'k'],
+                               'k' => ['n'],
+                               'l' => ['m'],
+                               'm' => ['g'],
+                               'n' => ['c', 'j', 'm']))
 
 .. code-block:: julia
 
-    infile = "web_graph_data.txt"
-    alphabet = "abcdefghijklmnopqrstuvwxyz"
-
-    n = 14 # total number of web pages (nodes)
-
-    # create adjacency matrix of links (Q[i, j] = 1 for link, 0 otherwise)
-    Q = fill(0, n, n)
-    edges = readlines(infile)
-    for edge in edges
-        from_node, to_node = [String(m.match) for m = eachmatch(r"\w", edge)] # String() required to go from Substring to String. r"\w" is a word character.
-        i = findfirst(isequal(from_node[1]), alphabet) # [1] required to go from String to Char
-        j = findfirst(isequal(to_node[1]), alphabet)
-        Q[i, j] = 1
-    end
+    nodes = keys(web_graph_data)
+    # create adjacency matrix of links (Q[i, j] = true for link, false otherwise)
+    Q = mapreduce(edges -> (nodes .∈ Ref(edges))', vcat, values(web_graph_data))
```

Comment:
Cool.  But *WAY* too tricky for introductory programmers.  I think the basic test is: if `reduce` and especially `mapreduce` and `vcat` or `hcat` are used, then there is no chance introductory users will understand it.  I think we just need to entirely hold off on functional programming patterns of this sort.

BTW, I think that in a future iteration of the lectures, we should add in a "functional programming and parallel computing" section where we can go through these sorts of programming patterns carefully, and explain why they are so essential to get parallel computing working well.

I really like the separation of the file loading from the creation of the graph is a major improvement, though.

Reply:
But I think it is worth leaving in that line as code, commented out, to students with a comment that says something to the effort of "functional programming style can be very concise, if you invest some time in learning it"

Reply:
It's neat code and some people love functional programming -- although I agree with @jlperla  that it  simple explicit loops will be better for the basic implementation.

Perhaps what's needed is a short lecture on functional programming in Julia that's flagged at the top as more advanced.  This would be a good example to use in that lecture.

Reply:
Also, in this specific case, my suspicion is that - even conditioning on following a lecture on functional programming, a more direct loop and iterator approach is better for the notes.

I generally like functional techniques for my own stuff, but fully understanding that code and convincing myself it is correct would take me minimum 5+ minutes of playing with MWE in the REPL.. And I would probably end up writing the loop version out of paranoia in a unit test.  So, even if I needed it to be functional to parallelize, I would probably have written the loop version as a check 

This is probably an example of where brevity does not bring clarity.